# CÂU 5: TỶ LỆ ÁP DỤNG KHUYẾN MÃI TRONG ĐƠN HÀNG

**1. Yêu cầu bài toán:**
Tính tỷ lệ phần trăm các dòng đơn hàng trong file `order_items.csv` có áp dụng mã giảm giá (tức là trường `promo_id` không bị trống/null).

**2. Phương pháp tiếp cận & Suy luận:**
Để tính được tỷ lệ phần trăm này, ta cần:
* **Bước 1:** Tải dữ liệu chi tiết đơn hàng từ file `order_items.csv`.
* **Bước 2:** Đếm tổng số lượng dòng dữ liệu có trong bảng.
* **Bước 3:** Đếm số lượng dòng mà cột `promo_id` chứa dữ liệu hợp lệ (không phải giá trị rỗng NaN).
* **Bước 4:** Lấy số lượng ở Bước 3 chia cho tổng số lượng ở Bước 2, rồi nhân với 100 để ra tỷ lệ phần trăm.

**3. Lý do chọn phương pháp:**
* **Sử dụng `notna()`:** Trong Pandas, dữ liệu trống được ký hiệu là `NaN` (Not a Number). Hàm `notna()` (hoặc `notnull()`) là cách tối ưu nhất để duyệt qua hàng triệu dòng và đánh dấu `True` cho những ô có chứa mã khuyến mãi, và `False` cho những ô bị trống.
* **Tính toán trực tiếp:** Việc dùng hàm `sum()` trên kết quả của `notna()` sẽ tự động cộng dồn các giá trị `True` (tương đương với 1), giúp ta đếm được số dòng có khuyến mãi với tốc độ cực nhanh.

In [1]:
# Import thư viện
import pandas as pd
from google.colab import drive

# Kết nối với Google Drive
drive.mount('/content/drive')

Mounted at /content/drive


**Thực hiện Bước 1: Tải dữ liệu**

Nạp dữ liệu chi tiết của từng dòng sản phẩm trong các đơn hàng.

In [2]:
# 1. Khai báo đường dẫn
file_path = '/content/drive/MyDrive/datathon-2026-round-1/order_items.csv'

# 2. Đọc dữ liệu
df_items = pd.read_csv(file_path)

# Hiển thị thử 5 dòng đầu để xem cột promo_id có dạng như thế nào
df_items[['order_id', 'product_id', 'promo_id']].head()

/tmp/ipykernel_9535/3838923976.py:5: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  df_items = pd.read_csv(file_path)


,order_id,product_id,promo_id
0,1,2400,NaN
1,2,609,NaN
2,3,396,NaN
3,4,635,NaN
4,6,1935,NaN


**Thực hiện Bước 2, 3 & 4: Tính toán tỷ lệ phần trăm**

Ta lấy số lượng dòng có mã khuyến mãi chia cho tổng số dòng của toàn bộ bảng.

In [3]:
# Tính tổng số dòng trong bảng
total_rows = len(df_items)

# Đếm số dòng có promo_id KHÔNG bị trống (not null)
promo_applied_rows = df_items['promo_id'].notna().sum()

# Tính tỷ lệ phần trăm
promo_percentage = (promo_applied_rows / total_rows) * 100

print(f"Tổng số dòng: {total_rows}")
print(f"Số dòng có áp dụng khuyến mãi: {promo_applied_rows}")
print(f"--> Tỷ lệ phần trăm có áp dụng khuyến mãi là: {promo_percentage:.2f}%")

Tổng số dòng: 714669
Số dòng có áp dụng khuyến mãi: 276316
--> Tỷ lệ phần trăm có áp dụng khuyến mãi là: 38.66%


**KẾT LUẬN CUỐI CÙNG:**
Kết quả tính toán cho thấy có 276.316 dòng áp dụng khuyến mãi trên tổng số 714.669 dòng. Tỷ lệ phần trăm tính được là **38.66%**, xấp xỉ **39%**.

Đối chiếu với các phương án của đề bài:

A) 12%

B) 25%

C) 39%

D) 54%

**=> Đáp án được chọn là: C**